File: package-tracker-backend/data_generator.py
"""
基于test.ipynb的数据生成逻辑
作者: 孙石，朱虹翱
"""

import random
import numpy as np
import uuid
from sklearn.cluster import KMeans
from typing import List, Dict, Tuple, Any

# 全局参数配置
parameters = {
    "station_num": 25,
    "center_num": 5,
    "packet_num": 10,
}

def data_gen() -> Dict[str, Any]:
    """
    生成站点、中心、边和包裹数据
    返回包含所有生成数据的字典
    """
    # Generate Stations
    station_pos = []
    # properties are defined here: throughput/tick, time_delay, money_cost
    station_prop_candidates = [
        (10, 2, 0.5), (15, 2, 0.6), (20, 1, 0.8), (25, 1, 0.9)]
    station_prop = []
    
    for i in range(parameters["station_num"]):
        # Map size is defined here, which is 100*100
        station_pos.append((random.randint(0, 100), random.randint(0, 100)))
        station_prop.append(
            station_prop_candidates[random.randint(0, len(station_prop_candidates)-1)])

    # Generate Centers by clustering
    kmeans = KMeans(n_clusters=parameters["center_num"], random_state=42, n_init=10)
    kmeans.fit(station_pos)
    station_labels = kmeans.predict(station_pos)
    center_pos = [(int(x[0]), int(x[1])) for x in kmeans.cluster_centers_]
    
    for i in range(len(center_pos)):
        while center_pos[i] in station_pos:
            # move slightly if center is overlapped with station
            print("Warning: Center moved")
            center_pos[i] = (center_pos[i][0] + 1, center_pos[i][1] + 1)
    
    # properties are defined here: throughput/tick, time_delay, money_cost
    center_prop_candidates = [
        (100, 2, 0.5), (150, 2, 0.5), (125, 1, 0.5), (175, 1, 0.5)]
    center_prop = []
    
    for i in range(parameters["center_num"]):
        center_prop.append(
            center_prop_candidates[random.randint(0, len(center_prop_candidates)-1)])

    # Generate Edges
    edges = []
    
    # 中心到中心的边（航线）
    for i in range(parameters["center_num"]):
        for j in range(parameters["center_num"]):
            if j > i:
                dist = np.linalg.norm(
                    np.array(center_pos[i]) - np.array(center_pos[j]))
                # src, dst, time_cost, money_cost
                edges.append((f"c{i}", f"c{j}", 0.25 * dist, 0.2 * dist))
                edges.append((f"c{j}", f"c{i}", 0.25 * dist, 0.2 * dist))

    # 中心到站点的边（高速公路）
    for i in range(parameters["center_num"]):
        for j in range(parameters["station_num"]):
            if station_labels[j] == i:
                dist = np.linalg.norm(
                    np.array(center_pos[i]) - np.array(station_pos[j]))
                edges.append((f"c{i}", f"s{j}", 0.6 * dist, 0.12 * dist))
                edges.append((f"s{j}", f"c{i}", 0.6 * dist, 0.12 * dist))

    # 站点到站点的边（道路）
    for i in range(parameters["station_num"]):
        for j in range(parameters["station_num"]):
            if i > j and (np.linalg.norm(np.array(station_pos[i]) - np.array(station_pos[j])) < 30):
                dist = np.linalg.norm(
                    np.array(station_pos[i]) - np.array(station_pos[j]))
                edges.append((f"s{i}", f"s{j}", 0.8 * dist, 0.07*dist))
                edges.append((f"s{j}", f"s{i}", 0.8 * dist, 0.07*dist))

    # 构建成本矩阵
    M = np.zeros((2*(parameters["center_num"]+parameters["station_num"]),
                  2*(parameters["center_num"]+parameters["station_num"])))
    for i in range(2*(parameters["center_num"]+parameters["station_num"])):
        for j in range(2*(parameters["center_num"]+parameters["station_num"])):
            M[i][j] = np.inf
    
    for i in range(parameters["center_num"]+parameters["station_num"]):
        M[2*i][2*i+1] = 0.01

    # 填充时间成本矩阵
    for i in range(parameters["center_num"]):
        for j in range(parameters["center_num"]):
            if j > i:
                M[2*i+1][2*j] = 0.25*np.linalg.norm(
                    np.array(center_pos[i]) - np.array(center_pos[j]))
                M[2*j+1][2*i] = M[2*i+1][2*j]

    for i in range(parameters["center_num"]):
        for j in range(parameters["station_num"]):
            if station_labels[j] == i:
                M[2*i+1][2*j+2*parameters["center_num"]] = 0.6*np.linalg.norm(
                    np.array(center_pos[i]) - np.array(station_pos[j]))
                M[2*j+2*parameters["center_num"]+1][2*i] = M[2*i+1][2*j+2*parameters["center_num"]]

    for i in range(parameters["station_num"]):
        for j in range(parameters["station_num"]):
            if i > j and (np.linalg.norm(np.array(station_pos[i]) - np.array(station_pos[j])) < 30):
                M[2*i+2*parameters["center_num"]+1][2*j+2*parameters["center_num"]] = 0.8*np.linalg.norm(
                    np.array(station_pos[i]) - np.array(station_pos[j]))
                M[2*j+2*parameters["center_num"]+1][2*i+2*parameters["center_num"]] = M[2*i+2*parameters["center_num"]+1][2*j+2*parameters["center_num"]]

    # 构建金钱成本矩阵
    N = np.zeros((2*(parameters["center_num"]+parameters["station_num"]),
                  2*(parameters["center_num"]+parameters["station_num"])))
    for i in range(2*(parameters["center_num"]+parameters["station_num"])):
        for j in range(2*(parameters["center_num"]+parameters["station_num"])):
            N[i][j] = np.inf

    for i in range(parameters["center_num"]):
        for j in range(parameters["center_num"]):
            if j > i:
                N[2*i+1][2*j] = 0.2*np.linalg.norm(
                    np.array(center_pos[i]) - np.array(center_pos[j]))
                N[2*j+1][2*i] = N[2*i+1][2*j]

    for i in range(parameters["center_num"]):
        for j in range(parameters["station_num"]):
            if station_labels[j] == i:
                N[2*i+1][2*j+2*parameters["center_num"]] = 0.12*np.linalg.norm(
                    np.array(center_pos[i]) - np.array(station_pos[j]))
                N[2*j+2*parameters["center_num"]+1][2*i] = N[2*i+1][2*j+2*parameters["center_num"]]

    for i in range(parameters["station_num"]):
        for j in range(parameters["station_num"]):
            if i > j and (np.linalg.norm(np.array(station_pos[i]) - np.array(station_pos[j])) < 30):
                N[2*i+2*parameters["center_num"]+1][2*j+2*parameters["center_num"]] = 0.07*np.linalg.norm(
                    np.array(station_pos[i]) - np.array(station_pos[j]))
                N[2*j+2*parameters["center_num"]+1][2*i+2*parameters["center_num"]] = N[2*i+2*parameters["center_num"]+1][2*j+2*parameters["center_num"]]

    # 添加节点处理成本
    for i in range(parameters["center_num"]):
        N[2*i][2*i+1] = center_prop[i][2]
    for i in range(parameters["station_num"]):
        N[2*i+2*parameters["center_num"]][2*i+2*parameters["center_num"]+1] = station_prop[i][2]

    # Generate Packets
    packets = []
    src_prob = np.random.random(parameters["station_num"])
    src_prob = src_prob / np.sum(src_prob)
    dst_prob = np.random.random(parameters["station_num"])
    dst_prob = dst_prob / np.sum(dst_prob)
    # Package categories are defined here: 0 for Regular, 1 for Express
    speed_prob = [0.7, 0.3]
    
    for i in range(parameters["packet_num"]):
        src = np.random.choice(parameters["station_num"], p=src_prob)
        dst = np.random.choice(parameters["station_num"], p=dst_prob)
        while dst == src:
            dst = np.random.choice(parameters["station_num"], p=dst_prob)
        category = np.random.choice(2, p=speed_prob)
        # Create time of the package, during 12 time ticks(hours)
        create_time = np.random.random() * 12
        packet_id = str(uuid.uuid4())
        packets.append((packet_id, create_time, f"s{src}", f"s{dst}", category))

    # Sort packets by create time
    packets.sort(key=lambda x: x[1])

    return {
        "station_pos": station_pos,
        "station_prop": station_prop,
        "center_pos": center_pos,
        "center_prop": center_prop,
        "edges": edges,
        "packets": packets,
        "station_labels": station_labels.tolist(),
        "time_cost_matrix": M,
        "money_cost_matrix": N,
        "parameters": parameters
    }

def format_data_for_api(data: Dict[str, Any]) -> Dict[str, Any]:
    """
    将生成的数据格式化为API友好的格式
    """
    # 格式化站点数据
    stations = []
    for i, (pos, prop) in enumerate(zip(data["station_pos"], data["station_prop"])):
        stations.append({
            "id": f"s{i}",
            "pos": list(pos),
            "throughput": prop[0],
            "delay": prop[1],
            "cost": prop[2],
            "type": "station"
        })

    # 格式化中心数据
    centers = []
    for i, (pos, prop) in enumerate(zip(data["center_pos"], data["center_prop"])):
        centers.append({
            "id": f"c{i}",
            "pos": list(pos),
            "throughput": prop[0],
            "delay": prop[1],
            "cost": prop[2],
            "type": "center"
        })

    # 格式化边数据
    edges = []
    for edge in data["edges"]:
        src, dst, time_cost, money_cost = edge
        # 确定边的类型
        edge_type = "road"  # 默认
        if src.startswith('c') and dst.startswith('c'):
            edge_type = "airline"
        elif (src.startswith('c') and dst.startswith('s')) or (src.startswith('s') and dst.startswith('c')):
            edge_type = "highway"
        
        edges.append({
            "src": src,
            "dst": dst,
            "timeCost": float(time_cost),
            "moneyCost": float(money_cost),
            "type": edge_type
        })

    # 格式化包裹数据
    packets = []
    for packet in data["packets"]:
        packet_id, create_time, src, dst, category = packet
        packets.append({
            "id": packet_id,
            "createTime": float(create_time),
            "src": src,
            "dst": dst,
            "category": int(category),
            "status": "created",
            "currentLocation": src,
            "history": [{
                "timestamp": float(create_time),
                "location": src,
                "action": "包裹已创建"
            }]
        })

    return {
        "stations": stations,
        "centers": centers,
        "edges": edges,
        "packets": packets,
        "parameters": data["parameters"],
        "timeCostMatrix": data["time_cost_matrix"].tolist(),
        "moneyCostMatrix": data["money_cost_matrix"].tolist()
    }

File: package-tracker-frontend/src/data/DataGenerator.js
// 模拟包裹追踪系统的数据生成和管理
// 基于您的 Python 代码逻辑

class DataGenerator {
  constructor() {
    this.parameters = {
      station_num: 25,
      center_num: 5,
      packet_num: 10,
    };
  }

  // 生成随机站点位置和属性
  generateStations() {
    const stations = [];
    const stationPropCandidates = [
      [10, 2, 0.5], [15, 2, 0.6], [20, 1, 0.8], [25, 1, 0.9]
    ];

    for (let i = 0; i < this.parameters.station_num; i++) {
      const pos = [
        Math.floor(Math.random() * 100),
        Math.floor(Math.random() * 100)
      ];
      const prop = stationPropCandidates[
        Math.floor(Math.random() * stationPropCandidates.length)
      ];
      
      stations.push({
        id: `s${i}`,
        pos: pos,
        throughput: prop[0],
        delay: prop[1],
        cost: prop[2],
        type: 'station'
      });
    }

    return stations;
  }

  // 使用简单聚类算法生成中心点
  generateCenters(stations) {
    const centers = [];
    const centerPropCandidates = [
      [100, 2, 0.5], [150, 2, 0.5], [125, 1, 0.5], [175, 1, 0.5]
    ];

    // 简单的 K-means 聚类模拟
    const clusterCenters = this.simpleKMeans(
      stations.map(s => s.pos), 
      this.parameters.center_num
    );

    for (let i = 0; i < this.parameters.center_num; i++) {
      const prop = centerPropCandidates[
        Math.floor(Math.random() * centerPropCandidates.length)
      ];
      
      centers.push({
        id: `c${i}`,
        pos: [
          Math.round(clusterCenters[i][0]),
          Math.round(clusterCenters[i][1])
        ],
        throughput: prop[0],
        delay: prop[1],
        cost: prop[2],
        type: 'center'
      });
    }

    return centers;
  }

  // 简单的 K-means 聚类算法
  simpleKMeans(points, k) {
    // 初始化中心点
    let centers = [];
    for (let i = 0; i < k; i++) {
      centers.push([
        Math.random() * 100,
        Math.random() * 100
      ]);
    }

    // 迭代优化
    for (let iter = 0; iter < 10; iter++) {
      const clusters = Array(k).fill().map(() => []);
      
      // 分配点到最近的中心
      points.forEach(point => {
        let minDist = Infinity;
        let closestCenter = 0;
        
        centers.forEach((center, i) => {
          const dist = Math.sqrt(
            Math.pow(point[0] - center[0], 2) + 
            Math.pow(point[1] - center[1], 2)
          );
          if (dist < minDist) {
            minDist = dist;
            closestCenter = i;
          }
        });
        
        clusters[closestCenter].push(point);
      });

      // 更新中心点
      centers = clusters.map(cluster => {
        if (cluster.length === 0) return centers[0];
        
        const sumX = cluster.reduce((sum, p) => sum + p[0], 0);
        const sumY = cluster.reduce((sum, p) => sum + p[1], 0);
        return [sumX / cluster.length, sumY / cluster.length];
      });
    }

    return centers;
  }

  // 生成边/路径
  generateEdges(stations, centers) {
    const edges = [];
    
    // 中心到中心的连接（航线）
    for (let i = 0; i < centers.length; i++) {
      for (let j = i + 1; j < centers.length; j++) {
        const dist = this.calculateDistance(centers[i].pos, centers[j].pos);
        const timeCost = 0.25 * dist;
        const moneyCost = 0.2 * dist;
        
        edges.push({
          src: centers[i].id,
          dst: centers[j].id,
          timeCost: timeCost,
          moneyCost: moneyCost,
          distance: dist,
          type: 'airline'
        });
        
        edges.push({
          src: centers[j].id,
          dst: centers[i].id,
          timeCost: timeCost,
          moneyCost: moneyCost,
          distance: dist,
          type: 'airline'
        });
      }
    }

    // 为每个站点分配最近的中心
    const stationLabels = stations.map(station => {
      let minDist = Infinity;
      let closestCenter = 0;
      
      centers.forEach((center, i) => {
        const dist = this.calculateDistance(station.pos, center.pos);
        if (dist < minDist) {
          minDist = dist;
          closestCenter = i;
        }
      });
      
      return closestCenter;
    });

    // 中心到站点的连接（高速公路）
    stations.forEach((station, i) => {
      const centerIndex = stationLabels[i];
      const center = centers[centerIndex];
      const dist = this.calculateDistance(station.pos, center.pos);
      const timeCost = 0.6 * dist;
      const moneyCost = 0.12 * dist;
      
      edges.push({
        src: center.id,
        dst: station.id,
        timeCost: timeCost,
        moneyCost: moneyCost,
        distance: dist,
        type: 'highway'
      });
      
      edges.push({
        src: station.id,
        dst: center.id,
        timeCost: timeCost,
        moneyCost: moneyCost,
        distance: dist,
        type: 'highway'
      });
    });

    // 站点到站点的连接（道路）- 仅连接距离小于30的站点
    for (let i = 0; i < stations.length; i++) {
      for (let j = i + 1; j < stations.length; j++) {
        const dist = this.calculateDistance(stations[i].pos, stations[j].pos);
        if (dist < 30) {
          const timeCost = 0.8 * dist;
          const moneyCost = 0.07 * dist;
          
          edges.push({
            src: stations[i].id,
            dst: stations[j].id,
            timeCost: timeCost,
            moneyCost: moneyCost,
            distance: dist,
            type: 'road'
          });
          
          edges.push({
            src: stations[j].id,
            dst: stations[i].id,
            timeCost: timeCost,
            moneyCost: moneyCost,
            distance: dist,
            type: 'road'
          });
        }
      }
    }

    return { edges, stationLabels };
  }

  // 计算两点间距离
  calculateDistance(pos1, pos2) {
    return Math.sqrt(
      Math.pow(pos1[0] - pos2[0], 2) + 
      Math.pow(pos1[1] - pos2[1], 2)
    );
  }

  // 生成包裹
  generatePackets(stations) {
    const packets = [];
    
    for (let i = 0; i < this.parameters.packet_num; i++) {
      const srcIndex = Math.floor(Math.random() * stations.length);
      let dstIndex = Math.floor(Math.random() * stations.length);
      
      // 确保源和目标不同
      while (dstIndex === srcIndex) {
        dstIndex = Math.floor(Math.random() * stations.length);
      }
      
      const category = Math.random() < 0.7 ? 0 : 1; // 70% 普通包裹，30% 快递
      const createTime = Math.random() * 12; // 12小时内创建
      
      packets.push({
        id: this.generateUUID(),
        createTime: createTime,
        src: stations[srcIndex].id,
        dst: stations[dstIndex].id,
        category: category, // 0: 标准, 1: 快递
        status: 'created',
        currentLocation: stations[srcIndex].id,
        history: [{
          timestamp: createTime,
          location: stations[srcIndex].id,
          action: 'Colis créé'
        }],
        path: []
      });
    }
    
    // 按创建时间排序
    packets.sort((a, b) => a.createTime - b.createTime);
    
    return packets;
  }

  // 生成UUID
  generateUUID() {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
      const r = Math.random() * 16 | 0;
      const v = c === 'x' ? r : (r & 0x3 | 0x8);
      return v.toString(16);
    });
  }

  // 生成完整的数据集
  generateData() {
    const stations = this.generateStations();
    const centers = this.generateCenters(stations);
    const { edges, stationLabels } = this.generateEdges(stations, centers);
    const packets = this.generatePackets(stations);

    return {
      stations,
      centers,
      edges,
      packets,
      stationLabels,
      parameters: this.parameters
    };
  }
}

export default DataGenerator;